### ISSUE TITLE
Mitigate CSRF and DNS Rebinding attacks

### ISSUE THREAD

**Issue by zelivans** [coreos/etcd#9353]
After explaining the issue in private with the security mailing list, we determined it is low enough in severity to make it a public discussion.

This issue is relevant only to etcd deployment on local networks with no authentication scheme set up. It may not be a common scenario but it affects anyone who work with etcd locally or on a local network and use it without authentication, which is the default. For the sake of demonstration, I'm using a scenario where etcd is deployed on localhost. This can be any LAN address though (the attacker would have to know the address as a perquisite, but localhost is pretty common). 

The first issue is with [CSRF](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)). An attacker can set up a website that tries to send a POST request to the etcd server and modify a key. Adding a key is done with PUT so it is theoretically safe (can't PUT from an HTML form or such) but POST allows creating in-order keys that an attacker can send. 

Example PoC:
```
<html>
    <body onload="document.forms[0].submit()">
        <h1>CSRF</h1>
        <form action="http://localhost:2379/v2/keys/bla" method="POST">
            <input type="hidden" name="value" value="MALICIOUS"/>
            <input type="submit" value="Send"/>
        </form>
    </body>
</html>
```

The second issue is with [DNS rebinding](https://en.wikipedia.org/wiki/DNS_rebinding). It essentially means an attacker can control his DNS records to direct to localhost, and trick the browser into sending requests to localhost (or any other address).

There are many resources on how this attack works. 

I've set up a live PoC at [35.205.198.70](http://35.205.198.70) (based on [taviso's work](https://bugs.chromium.org/p/project-zero/issues/detail?id=1447)). If the issue is unclear though please let me know and I will explain the attack in further details.

Success example:
![image](https://user-images.githubusercontent.com/15815884/36641064-b981eecc-1a32-11e8-9f6c-458e349fd889.png)

Whitelisting hostnames is a possible simple solution. See [taviso's comment](https://bugs.chromium.org/p/project-zero/issues/detail?id=1447#c2) on this or the [fix](https://github.com/transmission/transmission/pull/468) he sent to Transmission.

**Comment by gyuho** [comment]
We've added `--host-whitelist` flag. This will be released in 3.4.

Thanks for report!

**Comment by sidhax** [comment]
any cve assigned to these issues ?

**Comment by zelivans** [comment]
@sidhax I've attempted to assign CVE IDs through DWF but CoreOS is acquired by Red Hat so it was 
forwarded to them and I wasn't yet informed of anything from their side

**Comment by pedrohc** [comment]
Two CVEs were assigned for this:

CVE-2018-1098 etcd: Cross-site request forgery via crafted local POST forms
CVE-2018-1099 etcd: DNS rebinding vulnerability in etcd server

**Comment by zelivans** [comment]
Thanks @pedrohc

**Comment by zhsj** [comment]
Hi, does this affect stable release? and is there any plan to backport the security fix, especially 3.2.x?
